### PR TITLE
Team event queue for MemberJoin events.

### PR DESCRIPTION
**Pull request by fisx**
Fixes https://github.com/zinfra/backend-issues/issues/1409

### CODE REVIEW

**Comment by fisx** [comment]
> Apart from them, I am curious why someone would want to use the last thing. If they want to just ignore all the events until now, they can just create a new UUIDv1 for current timestamp and use that as a starting point.

I'm not sure, but I thought that there is a random part in a UUIDv1, not just the time stamp.  That would mean that you'd have to guess the randomness correctly (unlikely), or the id will not be found and the entire queue will be returned.

File: services/galley/src/Galley/API/Error.hs [hunk 1]
```diff
@@ -244,3 +244,9 @@ federationNotImplemented qualified =
     rendered = LT.intercalate ", " . toList . fmap (LT.fromStrict . renderMapping) $ qualified
     renderMapping IdMapping {idMappingLocal, idMappingGlobal} =
       idToText idMappingLocal <> " -> " <> renderQualifiedId idMappingGlobal
+
+getTeamNotificationsBadLast :: Error
+getTeamNotificationsBadLast = Error status400 "bad-query" "Last team notification can not be pulled with `size` or `since` params."
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/galley/src/Galley/API/Error.hs:249]: Makes sense, I'll do that.

File: services/galley/src/Galley/API/TeamNotifications.hs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  if onlyLast
+    then do
+      unless (isNothing since) $ throwM getTeamNotificationsBadLast
+      unless (isNothing size) $ throwM getTeamNotificationsBadLast
+      mqn <- DataTeamQueue.fetchLast tid
+      case mqn of
+        Just qn -> pure $ queuedNotificationList [qn] False Nothing
+        Nothing -> throwM getTeamNotificationsNotFound
+    else do
+      page <- DataTeamQueue.fetch tid since (fromMaybe (unsafeRange 1000) size)
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/galley/src/Galley/API/TeamNotifications.hs:75]: Can you point me to any place in particular?

File: services/galley/test/integration/TestHelpers.hs [hunk 3]
```diff
@@ -47,3 +47,8 @@ test s n h = testCase n runTest
     runTest = do
       setup <- s
       void . flip runReaderT setup . runTestM $ h >> assertClean
+
+purgeQueue :: TestM ()
+purgeQueue = do
+  awsEnv <- fromJust <$> view tsAwsEnv
+  void . liftIO $ Aws.execute awsEnv readAllUntilEmpty
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/galley/test/integration/TestHelpers.hs:54]: thanks!  i tried to find that but couldn't.

File: services/galley/src/Galley/API/Teams.hs [hunk 4]
```diff
@@ -763,11 +768,34 @@ addTeamMemberInternal tid origin originConn newMem memList = do
...
+    ::: JSON ->
+  Galley Response
+getTeamNotificationsH (zusr ::: sinceRaw ::: size ::: onlyLast ::: _) =
+  json <$> APITeamQueue.getTeamNotifications zusr since size onlyLast
+  where
+    since :: Maybe NotificationId
+    since = parseUUID =<< sinceRaw
+    --
+    parseUUID :: ByteString -> Maybe NotificationId
+    parseUUID = fmap Id . (UUID.fromASCIIBytes >=> isV1UUID)
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/galley/src/Galley/API/Teams.hs:794]: will do!

**Comment by akshaymankar** [comment]
> I'm not sure, but I thought that there is a random part in a UUIDv1, not just the time stamp. That would mean that you'd have to guess the randomness correctly (unlikely),

My understanding of UUIDv1 is there is a timestamp part and there is a MAC address part. Timestamp part is more significant than MAC address part. So, if I generate a UUID on my computer at a given moment and you generate a UUID on your computer after me, my UUID is guaranteed to be smaller than yours (assuming our clocks are in sync).

If my understanding is correct, if someone wants to ignore all past events, they could just generate a UUID and start querying the backend with that.

> or the id will not be found and the entire queue will be returned.

We shouldn't return the whole queue if the id is not found. From what I see in the code, in this case we'd just end up returning 1 extra notification. Probably, we shouldn't do that either.

File: services/galley/test/integration/API/Teams.hs [hunk 5]
```diff
@@ -416,6 +418,64 @@ testAddTeamMember = do
...
+    queue1 <- getTeamQueue owner (Just n1) Nothing False
+    queue2 <- getTeamQueue owner (Just n2) Nothing False
+    liftIO $ assertEqual "team queue: from 1" [mem1, mem2] (snd <$> queue1)
+    liftIO $ assertEqual "team queue: from 2" [mem2] (snd <$> queue2)
+
+  do
+    -- unknown 'NotificationId'
+    let Just n1 = Id <$> UUID.fromText "615c4e38-950d-11ea-b0fc-7b04ea9f81c0"
+    queue <- getTeamQueue owner (Just n1) Nothing False
+    liftIO $ assertEqual "team queue: from unknown" (snd <$> queue) [mem1, mem2]
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/galley/test/integration/API/Teams.hs:452]: yes, there is a test missing now.

**Comment by fisx** [comment]
@mheinzel, since you asked: 7f0bc00 

No particularly strong reasons, but that's what I could come up with.

File: services/galley/src/Galley/API/TeamNotifications.hs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+-- | This module is a work-around for the fact that we do not send notifications to all users
+-- of large teams any more.  For greeting bots to hear from new users, the 'MemberJoin' event
+-- is stored in a queue that every member of a team can pull, in a similar (but not identical)
+-- way as gundeck's @/notifications@ end-point, galley has a @/teams/notifications@ end-point
+-- where these events can be pulled.
+--
+-- The module is a clone of "Gundeck.Notification".
```
Comment (mheinzel) [inline services/galley/src/Galley/API/TeamNotifications.hs:26]:
```suggestion

-- | This module is a clone of "Gundeck.Notification".
```

1. the comment is now available in Haddock documentation
2. the empty line separates the license header from the comment, allowing our license header management tool to do its job

File: services/galley/test/integration/API/Teams.hs [hunk 7]
```diff
@@ -451,6 +453,18 @@ testTeamQueue = do
     queue <- getTeamQueue owner (Just n1) Nothing False
     liftIO $ assertEqual "team queue: from old unknown" (snd <$> queue) [mem1, mem2]
 
+  do
+    -- unknown younger 'NotificationId"
```
Comment (akshaymankar) [inline services/galley/test/integration/API/Teams.hs:457]:
```suggestion
    -- unknown younger 'NotificationId'
```

File: services/galley/schema/src/V41_TeamNotificationQueue.hs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+import Cassandra.Schema
+import Imports
+import Text.RawString.QQ
+
+migration :: Migration
+migration = Migration 40 "Create team notification queue" $ do
+  schema'
+    [r|
+        CREATE TABLE notifications (
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/galley/schema/src/V41_TeamNotificationQueue.hs]: dd9c86aa6d143eb8ac224f8a4aec82b8cac0be93

File: services/galley/src/Galley/API/TeamNotifications.hs (hunk 6 above)
Comment (fisx) [reply services/galley/src/Galley/API/TeamNotifications.hs]: I did the `This`, but the `|` is already up there in the haddocks that are folded away.

ecc053fc29f040c1b72b0efacfc827d89c0cb2d5

File: services/galley/test/integration/API/Util.hs [hunk 9]
```diff
@@ -811,6 +822,51 @@ deleteUser u = do
...
+getTeamQueue zusr msince msize onlyLast = do
+  parseEventList . responseJsonUnsafe
+    <$> ( getTeamQueue' zusr msince (fst <$> msize) onlyLast
+            <!! const 200 === statusCode
+        )
+  where
+    parseEventList :: QueuedNotificationList -> [(NotificationId, UserId)]
+    parseEventList qnl
+      | isJust msize && qnl ^. queuedHasMore /= (snd $ fromJust msize) =
+        error $ "expected has_more: " <> show (snd $ fromJust msize) <> "; but found: " <> show (qnl ^. queuedHasMore)
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/galley/test/integration/API/Util.hs:839]: I was hoping it makes the actual test more concise and thus readable.

**Pull request merged**